File: api/routers/feedback.py
from fastapi import APIRouter, Depends, HTTPException
from models.schemas import FeedbackRequest, FeedbackResponse
from services.auth import verify_supabase_jwt
from services.database import get_supabase

router = APIRouter()


@router.post("/feedback", response_model=FeedbackResponse)
async def feedback(
    body: FeedbackRequest,
    user: dict = Depends(verify_supabase_jwt),
):
    supabase = get_supabase()

    # evaluation がログインユーザーのキーに紐づいているか確認
    eval_check = (
        supabase.table("evaluations")
        .select("id, api_keys!inner(user_id)")
        .eq("id", str(body.evaluation_id))
        .eq("api_keys.user_id", user["id"])
        .maybe_single()
        .execute()
    )
    if not eval_check.data:
        raise HTTPException(status_code=404, detail="Evaluation not found")

    result = (
        supabase.table("feedbacks")
        .insert(
            {
                "evaluation_id": str(body.evaluation_id),
                "rating": body.rating,
                "comment": body.comment,
            }
        )
        .execute()
    )

    return FeedbackResponse(id=result.data[0]["id"])

File: api-worker/src/routes/feedback.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { HTTPException } from "hono/http-exception";
import type { AppVariables, Bindings } from "../types";
import { verifyJwt } from "../lib/auth";
import { FeedbackRequestSchema } from "../schemas";

export const feedbackRoute = new Hono<{
  Bindings: Bindings;
  Variables: AppVariables;
}>();

feedbackRoute.post(
  "/feedback",
  verifyJwt,
  zValidator("json", FeedbackRequestSchema),
  async (c) => {
    const body = c.req.valid("json");
    const user = c.get("user");

    // evaluation がログインユーザーのキーに紐づいているか確認
    const owner = await c.env.DB.prepare(
      `SELECT e.id FROM evaluations e
         JOIN api_keys k ON e.api_key_id = k.id
        WHERE e.id = ? AND k.user_id = ?
        LIMIT 1`,
    )
      .bind(body.evaluation_id, user.id)
      .first<{ id: string }>();

    if (!owner) {
      throw new HTTPException(404, { message: "Evaluation not found" });
    }

    const id = crypto.randomUUID();
    const result = await c.env.DB.prepare(
      `INSERT INTO feedbacks (id, evaluation_id, rating, comment)
       VALUES (?, ?, ?, ?)`,
    )
      .bind(id, body.evaluation_id, body.rating, body.comment ?? null)
      .run();

    if (!result.success) {
      throw new HTTPException(500, {
        message: `Failed to insert feedback: ${result.error ?? "unknown"}`,
      });
    }

    return c.json({ id, status: "ok" });
  },
);
